lib/registry/parsers.ts

```diff
@@ -0,0 +1,377 @@
+/**
+ * @license
+ * Copyright 2017 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// --- globals
+
+export const DEFAULT_INDEX_NAME = 'docker.io';
+// const DEFAULT_INDEX_URL = 'https://index.docker.io';
+export const DEFAULT_V2_REGISTRY = 'https://registry-1.docker.io';
+export const DEFAULT_LOGIN_SERVERNAME = 'https://index.docker.io/v1/';
+
+const VALID_NS = /^[a-z0-9._-]*$/;
+const VALID_REPO = /^[a-z0-9_/.-]*$/;
+
+const ParseAuth = /(\w+)\s+(.*)/; // -> scheme, params
+const Separators = /([",=])/;
+
+// --- exports
+
+export type ParsedIndex = {
+	name: string;
+	official: boolean;
+	scheme: string;
+};
+
+export type ParsedRepo = {
+	localName: string;
+	canonicalName: string;
+	remoteName: string;
+	indexUrl: string;
+	official: boolean;
+	indexName: string;
+};
+
+export type HeaderParams = {
+	[key: string]: string;
+};
+
+export type ParsedChallenge = {
+	scheme?: string;
+	params?: HeaderParams;
+};
+
+enum ParamParseState {
+	Token,
+	ExpectEquals,
+	ExpectValue,
+	HandleQuotedString,
+	EndQuotedString,
+	ExpectComma,
+}
+
+function splitStrOnce(str: string, separator: string) {
+	const index = str.indexOf(separator);
+	if (index < 0) {
+		return str;
+	}
+	return [str.substring(0, index), str.substring(index + 1)];
+}
+
+/**
+ * Parse a docker index name or index URL.
+ *
+ * Examples:
+ *      docker.io               (no scheme implies 'https')
+ *      index.docker.io         (normalized to docker.io)
+ *      https://docker.io
+ *      http://localhost:5000
+ *
+ */
+function parseIndex(arg: string | undefined): ParsedIndex {
+	const index: Partial<ParsedIndex> = {};
+
+	if (!arg || arg === DEFAULT_LOGIN_SERVERNAME) {
+		// Default index.
+		index.name = DEFAULT_INDEX_NAME;
+		index.official = true;
+		index.scheme = 'https';
+	} else {
+		// Optional protocol/scheme.
+		let indexName: string;
+
+		const protoSepIdx = arg.indexOf('://');
```

Comment:
A great point.  The problem we have here is that the index may not be a URL, or it might be.  We don't actually know until we get into it.  Trying to apply `new URL(...)` will croak if it isn't really a proper URL.

Reply:
Isn't that the idea though? Try to apply `new URL(...)` and catch the error if it falls over, then we know it's not a URL?

Reply:
not sure I understand.  There are valid values that are not URL's.  We need to parse all of them. 
> Isn't that the idea though? Try to apply new URL(...) and catch the error if it falls over, then we know it's not a URL?